Log management system

kdrewes opened the pull request:
#### Checklist

- [✓] run `npm run test` and `npm run benchmark`
- [✓] tests and/or benchmarks are included
- [✓] documentation is changed or added
- [✓] commit message and code follows the [Developer's Certification of Origin]

Hello, I've been recently working on customizing the log management system which originally derived from  #4955 and #4463.  Below I have made the following updates
    
    • Made adjustments to the ``` createRequestLog(customMessage) function```
    
    ```  const customMessage = 'incoming request'
      if (disableRequestLogging === false) {
        childLogger.info({ req: request }, createRequestLogMessage(customMessage))
      }
```
    • Updated documentation on Server.md
     • Added 'createRequestLogMessage' prototype to fastify.d.ts

Uzlopak reviewed (changes requested):
This PR doesnt make any sense. Now you can change the "incoming request" by doing alot of unnecessary stuff ?!

Uzlopak commented on docs/Reference/Server.md:
```diff
@@ -1,3 +1,6 @@
+Server.md
+
+
```
```suggestion
```

What is this?

kdrewes replied:
Used as a notation.

Uzlopak commented on docs/Reference/Server.md:
```diff
@@ -374,6 +378,24 @@ The possible values this property may have are:
+
++ Default: `false`
+
+When logging is permitted, the user can fully customize their 
```
Text is not matching the described functionality. default is not false, but basically `v => v`

kdrewes replied:
Ok I'll be sure to correct that thanks.

Uzlopak commented on docs/Reference/Server.md:
````diff
@@ -2027,3 +2050,6 @@ fastify.listen({ port: 3000 }, (err) => {
   }
 })
 ```
+
+
+
````
```suggestion
```

kdrewes replied:
Thank you I will be sure to remove it.

Uzlopak commented on fastify.js:
```diff
@@ -760,10 +760,13 @@ function fastify (options) {
+        childLogger.info({ req: request }, createRequestLogMessage(customMessage))
       }
 
+      console.log('FST_ERR_BAD_URL(path): ', FST_ERR_BAD_URL(path))
+
```
Ah yes, console.log an Error. Double performance bottleneck

```suggestion
```

kdrewes replied:
Ok I will make sure to review that later.  Thanks

Uzlopak commented on fastify.js:
```diff
@@ -760,10 +760,13 @@ function fastify (options) {
       const request = new Request(id, null, req, null, childLogger, onBadUrlContext)
       const reply = new Reply(res, request, childLogger)
 
+      const customMessage = 'incoming request'
```
instantiate a variable which is only needed in the if block?

kdrewes replied:
One of my biggest struggles is trying figure out where the customizable message will originate from.  Please let me know if you have any suggestions.  Thank you for your feedback.

Uzlopak commented on test/router-options.test.js:
```diff
@@ -171,7 +171,7 @@ test('Should honor frameworkErrors option - FST_ERR_BAD_URL', t => {
   )
 })
-
+//
```
```suggestion
```

kdrewes replied:
I'll be sure to remove.

Uzlopak commented on test/router-options.test.js:
```diff
@@ -217,7 +217,7 @@ test('Should supply Fastify request to the logger in frameworkErrors wrapper - F
   )
 })
-
+//
```
```suggestion
```

kdrewes replied:
in test/router-options.test.js I was temporarily removing tests to determine where the errors were.  So the // was used as a placeholder.

jsumners replied:
If the PR is a work in progress, please mark it as a draft. Otherwise, please remove these extraneous lines.

kdrewes replied:
Ok sounds good thanks.

Uzlopak reviewed (changes requested):
Please clarify what the purpose of this change is. 

Sry but this looks like some random stuff generated by an AI, e.g. copilot

jsumners reviewed (changes requested):
There's a good bit of work to go before this is ready. Please review my notes.

Eomm reviewed (commented):
While this feature may land, I would like to catch this opportunity to write down a thought.

We have the `disableRequestLogging` flag already but the user to replicate the same feature and change the log message should:

1. set the `disableRequestLogging`
2. add 3 hooks (onRequest, onResponse, onError)
3. eventually configure some serializers

I think we need to find a new API to let a plugin to set the server's configuration in order to avoid such syntax:

```
const companyFastifyLogger = myLogger()

const app = Fastify({ ...companyFastifyLogger.serverConfig })
app.register(companyFastifyLogger.plugin)
```

This would help other server's config too ([one example is the `schemaCompiler`](https://github.com/Eomm/joi-compiler?tab=readme-ov-file#usage)).

This PR shows that it is something necessary and we can build a more long term feature.

---

That said, the `disableRequestLogging` flag should allow a dev from achieving the same result, so my general opinion about the proposed option is that it is too specific. If there are not real blockers, I would keep the core slimmer

kdrewes commented on docs/Reference/Server.md:
```diff
@@ -1,3 +1,6 @@
+Server.md
+
+
 <h1 align="center">Fastify</h1>
```
Absolutely, I will make sure to revert this change, my apologies for the oversight.

kdrewes commented on docs/Reference/Server.md:
````diff
@@ -374,6 +378,24 @@ The possible values this property may have are:
   const fastify = require('fastify')({logger: customLogger});
   ```
 
+### `createRequestLogMessage`
+<a id="create-request-log-message"></a>
+
++ Default: `false`
+
+When logging is permitted, the user can fully customize their 
+log settings to suit their needs without relying on the default settings.  
+As a result, the user is no longer limited to the default layout.
+Please view the following example:
+
+```js
+  const customMessage = 'incoming request'
+
+  if (disableRequestLogging === false) {
+  childLogger.info({ req: request }, createRequestLogMessage(customMessage))
+  } else { childLogger.info({ req: request }, 'incoming request') }
+```
````
No problem I will make to revise the documentation.  Is there a specific format you would recommend I follow?

As for the documentation, the message I want to convey is that the user will have the ability to customize the log management system if disableRequestLogging is enabled.  The user will be able to customize this feature to fit their objective.  As for the code, this was a mistake that I should've corrected.  Originally I had it set up that way but then I was encountering an error so I had to change it.

kdrewes commented on docs/Reference/Server.md:
````diff
@@ -2027,3 +2050,6 @@ fastify.listen({ port: 3000 }, (err) => {
   }
 })
 ```
+
+
+
````
Ok sounds good will do.

kdrewes commented on fastify.js:
```diff
@@ -760,10 +760,13 @@ function fastify (options) {
       const reply = new Reply(res, request, childLogger)
 
+      const customMessage = 'incoming request'
       if (disableRequestLogging === false) {
-        childLogger.info({ req: request }, 'incoming request')
+        childLogger.info({ req: request }, createRequestLogMessage(customMessage))
       }
 
+      console.log('FST_ERR_BAD_URL(path): ', FST_ERR_BAD_URL(path))
```
So one of my biggest challenges was trying to determine where the customized log message originates from.  Is createRequestLogMessage the variable that would store this data?  Also, please let me know if theres anything I'm missing or not taking into account.

As for the code snippet thank you so much for your suggestion, I will make sure to experiment with it and also remove the console.log.

kdrewes commented:
> There's a good bit of work to go before this is ready. Please review my notes.

Thank you for all your help and feedback @jsumners.  I will continue to spend more time on this issue and apply all suggestions that were provided.

I also had a question I wanted to ask in regards to issue #4463

https://github.com/fastify/fastify/issues/4463

I noticed you originally made a suggestion to apply a factory function option to this feature.  I was wondering if you still think this is applicable and if so would you have any suggestions on getting started?  Thanks.

kdrewes commented:
> While this feature may land, I would like to catch this opportunity to write down a thought.
> 
> We have the `disableRequestLogging` flag already but the user to replicate the same feature and change the log message should:
> 
> 1. set the `disableRequestLogging`
> 2. add 3 hooks (onRequest, onResponse, onError)
> 3. eventually configure some serializers
> 
> I think we need to find a new API to let a plugin to set the server's configuration in order to avoid such syntax:
> 
> ```
> const companyFastifyLogger = myLogger()
> 
> const app = Fastify({ ...companyFastifyLogger.serverConfig })
> app.register(companyFastifyLogger.plugin)
> ```
> 
> This would help other server's config too ([one example is the `schemaCompiler`](https://github.com/Eomm/joi-compiler?tab=readme-ov-file#usage)).
> 
> This PR shows that it is something necessary and we can build a more long term feature.
> 
> That said, the `disableRequestLogging` flag should allow a dev from achieving the same result, so my general opinion about the proposed option is that it is too specific. If there are not real blockers I would work on them and I would keep the core slimmer

Thank you for all your help and suggestions @Eomm. 

After reading your post I agree; managing a customizable log management system seems like it could potentially be a laborious process compared to your suggestion of building an API which would set the server's configuration.  Even though my knowledge is quite limited, this does sound like a much more practical approach.  Therefore, I am open to any suggestions and ideas.  Please let me know if you have any questions or suggestions on how you guys would like to proceed.  Thank you.

jsumners commented:
> > There's a good bit of work to go before this is ready. Please review my notes.
> 
> Thank you for all your help and feedback @jsumners. I will continue to spend more time on this issue and apply all suggestions that were provided.
> 
> I also had a question I wanted to ask in regards to issue #4463
> 
> #4463
> 
> I noticed you originally made a suggestion to apply a factory function option to this feature. I was wondering if you still think this is applicable and if so would you have any suggestions on getting started? Thanks.

That was so long ago I don't have any idea what I was thinking. Still, it's really up to the person doing the work to determine the best solution. It sounds like @Eomm has some ideas he can help you with, though.

Uzlopak commented:
@kdrewes 

I hope you are well. I apologize for my snarky remarks and hope you will continue to contribute to fastify with this or other PR :). 

Do you need any further help or support to finish this PR?

Best Regards

Aras

kdrewes commented:
> @kdrewes
> 
> I hope you are well. I apologize for my snarky remarks and hope you will continue to contribute to fastify with this or other PR :).
> 
> Do you need any further help or support to finish this PR?
> 
> Best Regards
> 
> Aras

Hi @Uzlopak, I appreciate your apology but it's no worries at all.  I know your intention is only to help which is always appreciated.  For the time being I'm going to see if I can figure this out on my own but if I need any help (which is a big possibility) I will be sure to let you know. Thank you again for all your help.

Kyle

kdrewes commented:
> > > There's a good bit of work to go before this is ready. Please review my notes.
> > 
> > 
> > Thank you for all your help and feedback @jsumners. I will continue to spend more time on this issue and apply all suggestions that were provided.
> > I also had a question I wanted to ask in regards to issue #4463
> > #4463
> > I noticed you originally made a suggestion to apply a factory function option to this feature. I was wondering if you still think this is applicable and if so would you have any suggestions on getting started? Thanks.
> 
> That was so long ago I don't have any idea what I was thinking. Still, it's really up to the person doing the work to determine the best solution. It sounds like @Eomm has some ideas he can help you with, though.

Sounds good thank you @jsumners

kdrewes commented:
> While this feature may land, I would like to catch this opportunity to write down a thought.
> 
> We have the `disableRequestLogging` flag already but the user to replicate the same feature and change the log message should:
> 
> 1. set the `disableRequestLogging`
> 2. add 3 hooks (onRequest, onResponse, onError)
> 3. eventually configure some serializers
> 
> I think we need to find a new API to let a plugin to set the server's configuration in order to avoid such syntax:
> 
> ```
> const companyFastifyLogger = myLogger()
> 
> const app = Fastify({ ...companyFastifyLogger.serverConfig })
> app.register(companyFastifyLogger.plugin)
> ```
> 
> This would help other server's config too ([one example is the `schemaCompiler`](https://github.com/Eomm/joi-compiler?tab=readme-ov-file#usage)).
> 
> This PR shows that it is something necessary and we can build a more long term feature.
> 
> That said, the `disableRequestLogging` flag should allow a dev from achieving the same result, so my general opinion about the proposed option is that it is too specific. If there are not real blockers I would work on them and I would keep the core slimmer

Hi @Eomm, I hope all is well. Thank you again for your suggestion, I'd be interested in pursuing it further because I believe it would be more efficient and practical than my current task.  Before I get started I wanted to clarify a couple points:

• Would it be advisable to close this current PR and open a new one for this proposed task?
• Are there any additional features or characteristics you would wanted to be included in this plugin ?

Please let me know if you have any questions.  Thank you.

Kyle

kdrewes commented:
Hi @Eomm,

I hope all is well.  I have recently made some updates to the PR, which is still currently in draft.  Even though there are still changes that still need to be made, I wanted to share my progress and get your recommendations on how to proceed.  Below are the changes I’ve made so far:

* Removed all the unnecessary comments and notations from fastify.js, docs/References/Server.md and test/router-options.test.js

* Revised the documentation with docs/References/Server.md so the description will be more clear and concise

* Incorporated the syntax suggested by @jsumners used to customize the log management system depending on whether disableRequestLogging is enabled

 Please let me know if you have any questions.  Thank you.

jsumners commented on docs/Reference/Server.md:
````diff
@@ -374,6 +375,28 @@ The possible values this property may have are:
+Example:
+
+```js
+  if (disableRequestLogging === false) {
+        if (typeof createRequestLogMessage === 'function') {
+          childLogger.info({ req: request }, createRequestLogMessage(req, 'customMessage'))
+        } else {
+          childLogger.info({ req: request }, 'incoming request')
+        }
+      }
````
The indentation is not at all consistent in this block. I am also still unclear why this "example" is showing what looks like Fastify's internal implementation of the feature. It should be showing how to use it as user of Fastify.

kdrewes replied:
> The indentation is not at all consistent in this block.

Hi @jsumners, no problem, I'll make sure to re-adjust the indentation.

kdrewes replied:
>  I am also still unclear why this "example" is showing what looks like Fastify's internal implementation of the feature. It should be showing how to use it as user of Fastify.

Ok sounds good, I'll make sure to change that so it's from the user's perspective.

jsumners commented on docs/Reference/Server.md:
```diff
@@ -374,6 +375,28 @@ The possible values this property may have are:
+
++ Default: `true`
+
+When logging is enabled via `disableRequestLogging` parameter, users can 
+implement the `createRequestLogMessage` function to customize their log management 
+system.  As a result, this setting allows users to produce a customizable log message 
+catered to their projects' needs, circumventing the default format. 
```
```suggestion
When request logging is enabled, request logs can be customized by
supplying a `createRequestLogMessage(request, message)` function.
The function must be synchronous, and must not throw an error.
```

kdrewes replied:
Thank you @jsumners, I'll make sure incorporate this.

jsumners commented on docs/Reference/Server.md:
```diff
@@ -24,6 +24,7 @@ describes the properties available in that options object.
   - [`onProtoPoisoning`](#onprotopoisoning)
   - [`onConstructorPoisoning`](#onconstructorpoisoning)
   - [`logger`](#logger)
+  - [`createRequestLogMessage`](#createRequestLogMessage)
```
I'm not particularly fond of this name. I think something like `customRequestLog` would be better.

kdrewes replied:
> I'm not particularly fond of this name. I think something like `customRequestLog` would be better.

No problem I like that too!  I'll make sure to edit the variable name.

jsumners reviewed (changes requested):
Tests are missing.

kdrewes commented:
Hi guys,

I accidentally added this post a while ago to issue #4463 so I wanted to add it to the appropriate PR.

https://github.com/fastify/fastify/issues/4463#issuecomment-2871061817

I recently had some ideas I wanted to propose regarding the customization of the log management system. Below is the following:

1) Incorporate feature through onRequest Hook
The first approach I was considering was to incorporate a customizable log management system into our onRequest hook.
The user would now be able to make the necessary modifications and adjustments directly from the onRequest hook rather than from an arbitrary function located in the codebase

Pros:
• Less intrusive to the remainder of the codebase.
• Provides more scalability as it is now because it would be integrated with Fastify’s request response lifecycle
• Easier to facilitate because it is now incorporated with Fastify’s plugins

Cons:
• User would have to take on the burden of babysitting a hook
• User would have to oversee and cater to other features currently incorporated in the onRequest hook (I believe authentication, rate limiting and CORS)
• It could be more difficult to troubleshoot and debug because the user would have more features and functionalities to facilitate

2) Plugin Approach
The objective of the plug-in approach is to produce an object or plugin that is encapsulated by a childLogger object. This plugin/object will possess and inherit all the log management system properties it will need to create a customizable log management system. In addition, it can be reused and shared through several different Fastify projects.

Pros:
• Contains settings and properties directly adopted through the childLogger object. As a result, it has the capability to perform more functionalities and logic
• The whole entire Log Management system is now encapsulated through an object. Therefore, it makes it easier to perform functionalities. It’s also much more organized, scalable and maintainable than the previous approache
• Plugin is reusable

Cons:
• More challenging to configure because it requires calling and declaring many different features and properties in order to properly function. As a result, it will take more time to produce than the previous approache
• More difficult to debug and fix
• Has the tendency to be more cluttered and disorganized because of all the features and properties that hold it together
• Could potentially cause more overhead due to all the properties and utilities involved
• Requires more knowledge and expertise

Conclusion:
Personally, I believe both approaches are practical, however if I had to choose between the two I would have to select approach https://github.com/fastify/fastify/pull/3 for the following reasons:

• I believe having a plugin which encapsulates all the necessary childLogger properties would be a lot more practical because now we have our own entity which is capable of producing its own customizations. It could also be integrated in future projects or tasks that specifically involve utilizing the log management system for a variety of different purposes
• If we chose this option over the hook, then that means our plugin would not be subjected to solely being used during the request phase of the Request-Response lifecycle. Now why would this be in our best benefit? That's a good question that I don't have an answer to. However, at least it would provide us with more flexibility than if we were to incorporate our feature through an onRequest Hook
• We wouldn't have to take on the burden of monitoring and maintaining a hook

Thank you for taking the time to read my proposal. Please let me know if you have any concerns or objections regarding these approaches. Thank you.
